### PR TITLE
pkg/kubelet: recreate infra pod if the pod is changed

### PR DESCRIPTION
**Pull request by xiang90**
I was looking through the code and found this potential issue. I know the infra container would be recreated in next sync. I do not know if it is desired.

Also I did not handle the failure during the recreation as the old code.

/cc @dchen1107

### CODE REVIEW

File: pkg/kubelet/kubelet.go [hunk 1]
```diff
@@ -1167,6 +1159,20 @@ func (kl *Kubelet) syncPod(pod *api.BoundPod, dockerContainers dockertools.Docke
...
+		if podInfraContainer, found, _ := dockerContainers.FindPodContainer(podFullName, uid, dockertools.PodInfraContainerName); found {
+			if err := kl.killContainer(podInfraContainer); err != nil {
+				glog.V(1).Infof("Failed to kill pod infra container %q: %v", podInfraContainer.ID, err)
+			}
+		}
+		podInfraContainerID, err = kl.createPodInfraContainer(pod)
+		if err != nil {
+			glog.Errorf("Failed to recreate pod infra container: %v for pod %q", err, podFullName)
+		}
+		containersToKeep[podInfraContainerID] = empty{}
```
Comment (dchen1107) [inline pkg/kubelet/kubelet.go]: Is this necessary? At the beginning of Sync Loop, we should already put PodInfraContainerID to containersToKeep, right?

Comment (xiang90) [inline pkg/kubelet/kubelet.go]: `podInfraContainerID` is returned by createPodInfraContainer. It is called at line 1169.

File: pkg/kubelet/kubelet.go [hunk 2]
```diff
@@ -1088,6 +1089,7 @@ func (kl *Kubelet) syncPod(pod *api.BoundPod, dockerContainers dockertools.Docke
 				}
 				glog.Infof("pod %q container %q is unhealthy. Container will be killed and re-created.", podFullName, container.Name, live)
 			} else {
+				podChanged = true
```
Comment (thockin) [inline pkg/kubelet/kubelet.go]: Why would we restart the pod if just one container changed?

Comment (thockin) [inline pkg/kubelet/kubelet.go]: Ugh, there is an escape here - if a container adds/removes/changes a host port, we won't restart the network container.  I'm not too worried about that by itself, but are there any other aspects of a single container that might require the pod itself to be recreated?

Comment (xiang90) [inline pkg/kubelet/kubelet.go]: previously, we kill the infra pod container whenever there is a container change. This pull request does not change this behavior. The only change is to recreate the killed infra container after we finish the sync.

Comment (thockin) [inline pkg/kubelet/kubelet.go]: You're right.  That's awful

Comment (dchen1107) [inline pkg/kubelet/kubelet.go]: The only reason killing infra pod container when a container is changed is that we don't know if network related config is changed, such as port.

**Comment by dchen1107** [comment]
LGTM

**Pull request merged**